README.md
# youte: A command-line tool to retrieve and tidy YouTube metadata and comments from YouTube Data API

[![PyPI Latest Release](https://img.shields.io/pypi/v/youte)](https://pypi.org/project/youte/)
[![DOI](https://img.shields.io/badge/DOI-https%3A%2F%2Fdoi.org%2F10.25912%2FRDF__1684202566368-blue)](https://doi.org/10.25912/RDF_1684202566368)
[![Downloads](https://static.pepy.tech/personalized-badge/youte?period=month&units=international_system&left_color=black&right_color=orange&left_text=Monthly%20PyPI%20downloads)](https://pepy.tech/project/youte)
[![Licence](https://img.shields.io/github/license/QUT-Digital-Observatory/youte)](LICENCE)
![Supported Versions](https://img.shields.io/pypi/pyversions/youte)
![Development Status](https://img.shields.io/pypi/status/youte)

![](docs/images/youte_thumbnail.png)

## Changes from youte 1.3.0

Several major changes are made in youte 2.0. To better correspond with YouTube API endpoints and avoid confusion, some commands have been changed.

- `youte hydrate` is now broken down to `youte videos` and `youte channels`.
- `youte get-comments` is now `youte comments` and `youte replies`.
- `youte most-popular` is now `youte chart`.
- `youte get-related` is now `youte related-to`.

Furthermore:

- Resuming search is no longer available. Instead, you can set a limit on the number of search pages returned to avoid exhausting your API quota.
- All youte commands that retrieve data from YouTube API now won't print results to the shell, but store them in a specified json or jsonl file. This is to avoid clogging up the shell.
- You can now tidy data into a CSV or a flat JSON array.

Big thanks to @Lingomat (Mat Bettinson) for code review and suggestions.

## Installation

```bash
python -m pip install youte
```  

## YouTube API key  

To get data from YouTube API, you will need a YouTube API key. Follow YouTube [instructions](https://developers.google.com/youtube/v3/getting-started) to obtain a YouTube API key if you do not already have one.

## Configure API key (recommended)

You can save your API key in the youte config file for reuse. To do so, run:

```bash  
youte config add-key
```  

The interactive prompt will ask you to input your API key and name it. The name is used to identify the key, and can be anything you choose.

The prompt will also ask if you want to set the given key as default.

When running queries, if no API key or name is specified, `youte` will automatically use the default key.

### Manually set a key as default  

If you want to manually set an existing key as a default, run:  

```bash  
youte config set-default <name-of-existing-key>
```

Note that what is passed to this command is the _name_ of the API key, not the API key itself. It follows that the API key has to be first added to the config file using `youte config add-key`. If you use a name that has not been added to the config file, an error will be raised.

#### See the list of all keys  

To see the list of all keys, run:  

```bash  
youte config list-keys
```  

The default key, if there is one, will have an asterisk next to it.

#### Remove a key

To remove a stored key, run:

```bash
youte config remove <name-of-key>
```

#### About the config file  

youte's config file is stored in a central place whose exact location depends on the running operating system:  

- Linux/Unix: ~/.config/youte/   
- Mac OS X: ~/Library/Application Support/youte/
- Windows: C:\Users\\\<user>\\AppData\Roaming\youte

## search

Searching can be as simple as:

```bash
youte search <search-terms> --key <API-key> --outfile <name-of-file.json>
# OR
youte search <search-terms> --key <API-key> -o <name-of-file.json>
```

If you have a default key set up using `youte config`, then there is no need to specify an API key using `--key`.

This will return the maximum number of results pages (around 12-13) matching the search terms and store them in a JSON file. Unlike version 1.3, youte 2.0 does not print results to the terminal. Instead, `--outfile` is now a required option. <search-terms> and `--outfile` must be specified. 

In the search terms, you can also use the Boolean NOT (-) and OR (|) operators to exclude videos or to find videos that match one of several search terms. If the terms contain spaces, the entire search term value has to be wrapped in quotes.

Use the flag `--pretty` to pretty format the JSON output.

```bash
youte search <search-terms> --key <API-key> --outfile <name-of-file> --pretty
```

### Limit pages returned

Searching is very expensive in terms of API usage - a single results page uses up 100 points - 1% of your standard daily quota. Therefore, you can limit the maximum number of result pages returned, so that a search doesn't go on and exhaust your API quota.

```bash
youte search <search-terms> --max-pages 5
# OR
youte search <search-terms> -m 5
```

### Tidy data

Raw JSONs from YouTube API contain request metadata and nested fields. You can tidy these data into a CSV or a flat JSON using `--tidy-to`. The default format that youte will tidy raw JSON into will be CSV.

```bash
youte search <search-terms> --tidy-to <file.csv>
```

Specify `--format json` if you want to tidy raw data into an array of flat JSON objects.

```bash
youte search <search-terms> --tidy-to <file-name.json> --format json
```

`--tidy-to` option is available for all `youte` commands that retrieve data, and works the same way.

### Advanced search

There are multiple filters to refine your search. A full list of these are provided below:

``` {.bash .no-copy}
Options:
  --type TEXT                     Type of resource to search for  [default:
                                  video]
  --order [date|rating|relevance|title|videoCount|viewCount]
                                  Sort results  [default: date]
  --safe-search [none|moderate|strict]
                                  Include or exclude restricted content
                                  [default: none]
  --lang TEXT                     Return results most relevant to a language
                                  (ISO 639-1 two-letter code)
  --region TEXT                   Return videos viewable in the specified
                                  country (ISO 3166-1 alpha-2 code)  [default:
                                  US]
  --video-duration [any|long|medium|short]
                                  Include videos of a certain duration
  --channel-type [any|show]       Restrict search to a particular type of
                                  channel
  --video-type [any|episode|movie]
                                  Search a particular type of videos
  --caption [any|closedCaption|none]
                                  Filter videos based on if they have captions
  --definition, --video-definition [any|high|standard]
                                  Include videos by definition
  --dimension, --video-dimension [any|2d|3d]
                                  Search 2D or 3D videos
  --embeddable, --video-embeddable [any|true]
                                  Search only embeddable videos
  --license, --video-license [any|creativeCommon|youtube]
                                  Include videos with a certain license
  --location FLOAT...             Lat and long coordinates to restrict search
                                  to. --radius must be specified
  --radius TEXT                   Define the geographic area to restrict
                                  search. Must be a number with a unit
```

#### Restrict by date range

The `--from` and `--to` options allow you to restrict your search to a specific period. The input values have to be in ISO format (YYYY-MM-DD). Currently, all dates and times in youte are in UTC.

#### Restrict by type

The `--type` option specifies the type of results returned, which by default is videos. The accepted values are `channel`, `playlist`, `video`, or a combination of these three. If more than one type is specified, separate each by a comma.

```shell
youte search <search-terms> --limit 5 --type playlist,video
```

#### Restrict by language and region

The `--lang` returns results most relevant to a language. Not all results will be in the specified language: results in other languages will still be returned if they are highly relevant to the search query term. To specify the language, use [ISO 639-1 two letter code](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes), except that you should use the values `zh-Hans` for simplified Chinese and `zh-Hant` for traditional Chinese.

The `--region` returns results viewable in a region. To specify the region, use [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2). Note that this option does *not* filter videos uploaded in that region, but rather videos that can be _viewed_ in that region.

The `--location` and `--radius` options define a circular geographic area to filter videos that specify, in their metadata, a location within this area. This is *not* a robust and reliable way to geolocate YouTube videos, and hence should be used with care.

-  `--location` takes in 2 values - the latitude/longitude coordinates that represent the centre of the area
-  `--radius` specifies the maximum distance that the location associated with a video can be from that point for the video to still be included in the search results. It must be a number followed by a unit. Valid units are `m`, `km`, `ft`, and `mi`. For example, `1500m`, `5km`, `10000ft`, and `0.75mi`.

Both `--location` and `--radius` have to be specified if they are to be used, otherwise an API error will be thrown.

### Sort results

The `--order` option specifies how results will be sorted. The following values are accepted:

- `date`: Resources are sorted in reverse chronological order based on the date they were created (default value).
- `rating`: Resources are sorted from highest to lowest rating.
- `relevance`: Resources are sorted based on their relevance to the search query.
- `title` – Resources are sorted alphabetically by title.
- `videoCount` – Channels are sorted in descending order of their number of uploaded videos.
- `viewCount` – Resources are sorted from highest to lowest number of views. For live broadcasts, videos are sorted by number of concurrent viewers while the broadcasts are ongoing.

## videos

`youte videos` takes in one or multiple video IDs and retrieve all public metadata for those videos. This complements results returned from `youte search`, as they contain only limited information.

```shell

CHANGELOG.md
# Changelog

<!-- insertion marker -->
## Unreleased

<small>[Compare with latest](https://github.com/QUT-Digital-Observatory/youte/compare/v2.2.2...HEAD)</small>

<!-- insertion marker -->
## [v2.2.2](https://github.com/QUT-Digital-Observatory/youte/releases/tag/v2.2.2) - 2023-05-21

<small>[Compare with v2.2.1](https://github.com/QUT-Digital-Observatory/youte/compare/v2.2.1...v2.2.2)</small>

### Bug Fixes

- flattening json to tidy to csv should not flatten list ([bf272aa](https://github.com/QUT-Digital-Observatory/youte/commit/bf272aa2460820528ba3d04bc7ac219b701dfc71) by Boyd Nguyen).

## [v2.2.1](https://github.com/QUT-Digital-Observatory/youte/releases/tag/v2.2.1) - 2023-05-16

<small>[Compare with v2.2.0](https://github.com/QUT-Digital-Observatory/youte/compare/v2.2.0...v2.2.1)</small>

### Bug Fixes

- pre-existing types' subscriptability don't work in 3.8 ([0242454](https://github.com/QUT-Digital-Observatory/youte/commit/024245498538eee0fdbd6c4de7246c6a1c524757) by Boyd Nguyen).

## [v2.2.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/v2.2.0) - 2023-05-16

<small>[Compare with v2.1.0](https://github.com/QUT-Digital-Observatory/youte/compare/v2.1.0...v2.2.0)</small>

### Features

- include metadata in parsers ([d5d79fe](https://github.com/QUT-Digital-Observatory/youte/commit/d5d79fe44c275facd9cfb0897272c38cde9ba7d8) by Boyd Nguyen).
- add _youte metadata to returned output ([d87b97f](https://github.com/QUT-Digital-Observatory/youte/commit/d87b97fd2bc9f6092ed3bc4c55dc8d229ae4283c) by Boyd Nguyen).

### Bug Fixes

- fix None outdb when file doesn't exist ([802ab8f](https://github.com/QUT-Digital-Observatory/youte/commit/802ab8f2e97af9f161175bd2f6f9e41d74288aa8) by Boyd Nguyen).
- fix KeyError in parse_video + parse_videos ([6fbbb70](https://github.com/QUT-Digital-Observatory/youte/commit/6fbbb70f320e58c4aa85c2700cecaf0bd36f91da) by Boyd Nguyen).

## [v2.1.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/v2.1.0) - 2023-04-14

<small>[Compare with v2.0.2](https://github.com/QUT-Digital-Observatory/youte/compare/v2.0.2...v2.1.0)</small>

### Features

- new 'full-archive' command ([dbaef5c](https://github.com/QUT-Digital-Observatory/youte/commit/dbaef5cf5ab9acf2b5c4eafa0b948f7f34c779ae) by Boyd Nguyen).

### Bug Fixes

- improve logging ([66a7b68](https://github.com/QUT-Digital-Observatory/youte/commit/66a7b68a4e77c34a05e0f4208fe62a00b9bd3182) by Boyd Nguyen).
- use dict.get() to allow for nullable fields when parsing JSON ([3a2e8c9](https://github.com/QUT-Digital-Observatory/youte/commit/3a2e8c9db35639b31c1911da8b3d4b6905725802) by Boyd Nguyen).
- avoid inputting error messages to output file ([4a186dc](https://github.com/QUT-Digital-Observatory/youte/commit/4a186dc2b91f2d859439e773dd1a2ec62c1e4af6) by Boyd Nguyen).
- refine resource schema + make optional fields optional ([821bca9](https://github.com/QUT-Digital-Observatory/youte/commit/821bca98e1c237bffe90cb3f3356cefc74ab8987) by Boyd Nguyen).
- fix KeyError in parse_video + parse_videos ([05ade72](https://github.com/QUT-Digital-Observatory/youte/commit/05ade72ddd7ce644b573069434db275fac105f98) by Boyd Nguyen).

## [v2.0.2](https://github.com/QUT-Digital-Observatory/youte/releases/tag/v2.0.2) - 2023-03-31

<small>[Compare with 2.0.1](https://github.com/QUT-Digital-Observatory/youte/compare/2.0.1...v2.0.2)</small>

### Bug Fixes

- fix KeyError in parse_video + parse_videos ([8e9d4fe](https://github.com/QUT-Digital-Observatory/youte/commit/8e9d4fe2385431e8fbb28bc94c1c27578cde4b2b) by Boyd Nguyen).

## [2.0.1](https://github.com/QUT-Digital-Observatory/youte/releases/tag/2.0.1) - 2023-03-20

<small>[Compare with 2.0.0](https://github.com/QUT-Digital-Observatory/youte/compare/2.0.0...2.0.1)</small>

## [2.0.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/2.0.0) - 2023-03-16

<small>[Compare with 1.3.0](https://github.com/QUT-Digital-Observatory/youte/compare/1.3.0...2.0.0)</small>

## [1.3.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/1.3.0) - 2023-02-24

<small>[Compare with 1.2.0](https://github.com/QUT-Digital-Observatory/youte/compare/1.2.0...1.3.0)</small>

## [1.2.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/1.2.0) - 2023-02-07

<small>[Compare with 1.1.0](https://github.com/QUT-Digital-Observatory/youte/compare/1.1.0...1.2.0)</small>

## [1.1.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/1.1.0) - 2022-12-06

<small>[Compare with 1.0.1](https://github.com/QUT-Digital-Observatory/youte/compare/1.0.1...1.1.0)</small>

## [1.0.1](https://github.com/QUT-Digital-Observatory/youte/releases/tag/1.0.1) - 2022-12-01

<small>[Compare with 1.0.0](https://github.com/QUT-Digital-Observatory/youte/compare/1.0.0...1.0.1)</small>

## [1.0.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/1.0.0) - 2022-11-23

<small>[Compare with 0.1.1](https://github.com/QUT-Digital-Observatory/youte/compare/0.1.1...1.0.0)</small>

## [0.1.1](https://github.com/QUT-Digital-Observatory/youte/releases/tag/0.1.1) - 2022-10-10

<small>[Compare with 0.1.0](https://github.com/QUT-Digital-Observatory/youte/compare/0.1.0...0.1.1)</small>

## [0.1.0](https://github.com/QUT-Digital-Observatory/youte/releases/tag/0.1.0) - 2022-09-21


src/youte/__main__.py
"""
Copyright: Digital Observatory 2022 <digitalobservatory@qut.edu.au>
Author: Boyd Nguyen <thaihoang.nguyen@qut.edu.au>
"""
import json
import logging
import os
import sys
from json.decoder import JSONDecodeError
from pathlib import Path
from typing import List, Sequence, Union

import click

from youte import tidier
from youte.collector import Youte, _get_history_path
from youte.config import YouteConfig
from youte.exceptions import StopCollector, ValueAlreadyExists
from youte.utilities import validate_date_string

# Logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

file_handler = logging.FileHandler("youte.log")
file_handler.setLevel(logging.INFO)
file_formatter = logging.Formatter(
    "%(asctime)s - %(module)s: %(message)s (%(levelname)s)"
)
file_handler.setFormatter(file_formatter)

logger.addHandler(file_handler)


def _validate_date(ctx, param, value):
    if value:
        if validate_date_string(value):
            return value
        else:
            raise click.BadParameter("Date not in correct format (YYYY-MM-DD)")


# CLI argument set up:
@click.group()
@click.version_option()
def youte():
    """
    Utility to collect and tidy YouTube meta-data and comments via YouTube API.
    Run `youte --help` to get started.
    """
    pass


@youte.command()
@click.argument("query", required=True)
@click.argument("output", type=click.File(mode="w"), required=False)
@click.option(
    "--from", "from_", help="Start date (YYYY-MM-DD)", callback=_validate_date
)
@click.option("--to", help="End date (YYYY-MM-DD)", callback=_validate_date)
@click.option(
    "--type",
    "type_",
    default="video",
    help="Type of resource to search for",
    show_default=True,
)
@click.option("--name", help="Specify an API key name added to youte config")
@click.option("--key", help="Specify a YouTube API key")
@click.option(
    "--order",
    type=click.Choice(
        ["date", "rating", "relevance", "title", "videoCount", "viewCount"],
        case_sensitive=False,
    ),
    help="Sort results",
    show_default=True,
    default="date",
)
@click.option(
    "--safe-search",
    type=click.Choice(["none", "moderate", "strict"], case_sensitive=False),
    help="Include or exclude restricted content",
    default="none",
    show_default=True,
)
@click.option(
    "--video-duration",
    type=click.Choice(["any", "long", "medium", "short"]),
    help="Include videos of a certain duration",
)
@click.option(
    "--channel-type",
    type=click.Choice(["any", "show"]),
    help="Restrict search to a particular type of channel",
)
@click.option(
    "--video-type",
    type=click.Choice(["any", "episode", "movie"]),
    help="Search a particular type of videos",
)
@click.option(
    "--caption",
    type=click.Choice(["any", "closedCaption", "none"]),
    help="Filter videos based on if they have captions",
)
@click.option(
    "--definition",
    "--video-definition",
    "video_definition",
    type=click.Choice(["any", "high", "standard"]),
    help="Include videos by definition",
)
@click.option(
    "--dimension",
    "--video-dimension",
    "video_dimension",
    type=click.Choice(["any", "2d", "3d"]),
    help="Search 2D or 3D videos",
)
@click.option(
    "--embeddable",
    "--video-embeddable",
    "video_embeddable",
    type=click.Choice(["any", "true"]),
    help="Search only embeddable videos",
)
@click.option(
    "--license",
    "--video-license",
    "video_license",
    type=click.Choice(["any", "creativeCommon", "youtube"]),
    help="Include videos with a certain license",
)
@click.option(
    "--max-results",
    type=click.IntRange(0, 50),
    help="Maximum number of results returned per page",
    default=50,
    show_default=True,
)
@click.option(
    "--limit",
    "-l",
    type=click.IntRange(1, 13),
    help="Maximum number of result pages to retrieve",
)
@click.option("--resume", help="Resume progress from this file")
@click.option("--to-csv", type=click.Path(), help="Tidy data to CSV file")
def search(
    query: str,
    output: str,
    from_: str,
    to: str,
    name: str,
    key: str,
    order: str,
    video_duration: str,
    type_: str,
    channel_type: str,
    video_type: str,
    caption: str,
    video_definition: str,
    video_dimension: str,
    video_embeddable: str,
    video_license: str,
    safe_search: str,
    resume: str,
    to_csv: str,
    limit: int,
    max_results: int,
) -> None:
    """Do a YouTube search

    QUERY: search query

    OUTPUT: name of json file to store output data
    """
    api_key = key if key else _get_api_key(name=name)
    search_collector = _set_up_collector(api_key=api_key)

    params = {
        "part": "snippet",
        "q": query,
        "maxResults": max_results,
        "order": order,
        "safeSearch": safe_search,
        "videoDuration": video_duration,
        "type": type_,
        "channelType": channel_type,
        "videoCaption": caption,
        "videoDefinition": video_definition,
        "videoDimension": video_dimension,
        "videoEmbeddable": video_embeddable,
        "videoLicense": video_license,
        "videoType": video_type
    }

    if from_:
        params["publishedAfter"] = from_
    if to:
        params["publishedBefore"] = to

    try:
        if resume:
            if not _get_history_path(resume).exists():
                raise click.BadParameter("No such history file found")

        results = search_collector.search(
            save_progress_to=resume, limit=limit, **params
        )

        click.echo(params)

        for result in results:
            click.echo(json.dumps(result), file=output)

            if to_csv:
                items = result["items"]
                tidier.tidy_to_csv(items=items, output=to_csv, resource_kind="search")

        if output:
            click.echo(f"Results are stored in {output.name}")

    except StopCollector:
        _prompt_save_progress(search_collector.history_file)


@youte.command()
@click.argument("items", nargs=-1, required=False)
@click.option("-o", "--output", type=click.File(mode="w"), required=False)
@click.option("-f", "--file-path", help="Get IDs from file", default=None)
@click.option(
    "-t",
    "--by-thread",
    "by_parent",
    help="Get all replies to a parent comment",
    is_flag=True,
)
@click.option("-v", "--by-video", help="Get all comments for a video ID", is_flag=True)
@click.option("--name", help="Specify an API key name added to youte config")
@click.option("--key", help="Specify a YouTube API key")
@click.option("--to-csv", type=click.Path(), help="Tidy data to CSV file")
def get_comments(
    items: Sequence[str],
    output: str,
    name: str,
    key: str,
    file_path: str,
    by_video: bool,
    by_parent: bool,
    to_csv,
) -> None:
    """
    Get YouTube comments by video IDs or thread IDs

    The IDs should all belong to one type, i.e. either video or comment thread.
    You cannot mix both video AND comment thread IDs in one command.

    OUTPUT: name of JSON file to store output

    ITEMS: ID(s) of item as provided by YouTube
    """
    api_key = key if key else _get_api_key(name=name)
    collector = _set_up_collector(api_key=api_key)

    ids = _get_ids(string=items, file=file_path)

    if by_video and by_parent:
        click.secho(
            "Both video and parent flags are on. Only one is allowed.",
            fg="red",
            bold=True,
        )
        sys.exit(1)

    item_type = "comments" if by_parent else "comment_threads"

    if by_video:
        by = "video"
    elif by_parent:
        by = "parent"
    else:
        by = None

    results = collector.list_items(item_type=item_type, ids=ids, by=by)

    for result in results:
        click.echo(json.dumps(result), file=output)

        if to_csv:
            items = result["items"]
            tidier.tidy_to_csv(items=items, output=to_csv, resource_kind="comments")

    if output:
        click.echo(f"Results are stored in {output.name}")


@youte.command()
@click.argument("items", nargs=-1, required=False)
@click.option("-o", "--output", type=click.File(mode="w"), required=False)
@click.option("-f", "--file-path", help="Get IDs from file", default=None)
@click.option(
    "--kind",
    type=click.Choice(["videos", "channels", "comments"], case_sensitive=False),
    help="Sort results",
    show_default=True,
    default="videos",
)
@click.option("--name", help="Specify an API key name added to youte config")
@click.option("--key", help="Specify a YouTube API key")
@click.option("--to-csv", type=click.Path(), help="Tidy data to CSV file")
def hydrate(
    items: Sequence[str],
    output: str,
    kind: str,
    file_path: str,
    name: str,
    key: str,
    to_csv,
) -> Union[str, None]:
    """Hydrate YouTube resource IDs

    Get all metadata for a list of resource IDs.
    By default, the function hydrates video IDs.

    The IDs should all belong to one type, i.e. either video, channel, or comment.
    For example, you cannot mix both video AND channel IDs in one command.

    OUTPUT: name of JSON file to store output

    ITEMS: ID(s) of items as provided by YouTube
    """
    api_key = key if key else _get_api_key(name=name)
    collector = _set_up_collector(api_key=api_key)

    ids = _get_ids(string=items, file=file_path)

    results = collector.list_items(item_type=kind, ids=ids)

    for result in results:
        click.echo(json.dumps(result), file=output)

        if to_csv:
            items = result["items"]
            tidier.tidy_to_csv(items=items, output=to_csv, resource_kind=kind)

    if output:
        click.echo(f"Results are stored in {output.name}")


@youte.command()
@click.argument("items", nargs=-1, required=False)
@click.option(
    "-f", "--file-path", type=click.Path(), help="Get IDs from file", default=None
)
@click.option(
    "-o",
    "--output",
    type=click.File(mode="w"),
    help="Name of JSON file to store output",
)
@click.option(
    "--safe-search",
    type=click.Choice(["none", "moderate", "strict"], case_sensitive=False),
    help="Include or exclude restricted content",
    default="none",
    show_default=True,
)
@click.option("--name", help="Specify an API key name added to youte config")
@click.option("--key", help="Specify a YouTube API key")
@click.option(
    "--max-results",
    type=click.IntRange(0, 50),
    help="Maximum number of results returned per page",
    default=50,
    show_default=True,
)
@click.option("--to-csv", type=click.Path(), help="Tidy data to CSV file")
def get_related(
    items: Sequence[str],
    output: str,
    safe_search: str,
    to_csv: str,
    file_path: str,
    name: str,
    key: str,
    max_results: int,
):
    """Get videos related to a video

    ITEMS: ID(s) of videos as provided by YouTube
    """
    api_key = key if key else _get_api_key(name=name)
    collector = _set_up_collector(api_key=api_key)

    ids = _get_ids(string=items, file=file_path)

    for id_ in ids:
        params = {
            "part": "snippet",
            "maxResults": max_results,
            "safeSearch": safe_search,
            "type": "video",
            "relatedToVideoId": id_,
        }

        results = collector.search(**params)

        click.echo(params)

        for result in results:
            click.echo(json.dumps(result), file=output)

            if to_csv:
                items = result["items"]
                tidier.tidy_to_csv(items=items, output=to_csv, resource_kind="search")

        if output:
            click.echo(f"Results are stored in {output.name}")


@youte.command()
@click.argument("output", type=click.File(mode="w"), required=False)
@click.option(
    "-r", "--region-code", help="ISO 3166-1 alpha-2 country codes to retrieve videos"
)
@click.option("--name", help="Specify an API key name added to youte config")
@click.option("--key", help="Specify a YouTube API key")
@click.option("--to-csv", type=click.Path(), help="Tidy data to CSV file")
def most_popular(region_code, output, name, key, to_csv):
    """Return the most popular videos for a region and video category

    By default, if nothing is else is provided, the command retrieves the most
    popular videos in the US.

    OUTPUT: name of JSON file to store results
    """

    api_key = key if key else _get_api_key(name=name)
    collector = _set_up_collector(api_key=api_key)

    results = collector.list_most_popular(region_code=region_code)

    for result in results:
        click.echo(json.dumps(result), file=output)

        if to_csv:
            items = result["items"]
            tidier.tidy_to_csv(items=items, output=to_csv, resource_kind="videos")

    if output:
        click.echo(f"Results are stored in {output.name}")


@youte.command()
@click.argument("infile", type=click.Path())
@click.option(
    "-o", "--output", type=click.File(mode="w"), help="Output text file to store IDs in"
)
def dehydrate(infile, output: str) -> None:
    """Extract an ID list from a file of YouTube resources

    INFILE: JSON file of YouTube resources
    """
    try:
        items = tidier.get_items(infile)
        ids = tidier.get_id(items)
        for id_ in ids:
            click.echo(id_, file=output)
    except JSONDecodeError:
        raise click.BadParameter("File is not JSON.")


@youte.command()
@click.argument("filepath", nargs=-1, type=click.Path(), required=True)
@click.argument("output", type=click.Path(), required=True)
def tidy(filepath, output):
    """Tidy raw JSON response into relational SQLite databases"""
    for file in filepath:
        tidier.master_tidy(filepath=file, output=output)


@youte.command()
def list_history():
    """List resumable history files"""
    if not os.path.exists(".youte.history"):
        click.echo("No history file found")

    files = os.listdir(".youte.history")
    if not files:
        click.echo("No history file found")
    else:
        for file in files:
            click.echo(file.removesuffix(".db"))


@youte.group()
def config():
    """
    Set up API keys to access data from YouTube API
    """


@config.command()
def add_key():
    """
    Add YouTube API key
    """
    click.echo()
    click.secho("Welcome to youte!", fg="green", bold=True)
    click.echo("Configuring your profile...")
    click.echo()
    click.echo("To get started, an API key is required to get data from Youtube API.")
    click.echo()
    click.echo("To obtain an API key, follow the steps in")
    click.echo("https://developers.google.com/youtube/v3/getting-started")
    click.echo()
    click.echo("Once you have an API key, run `youte configure add-key` to start.")

    click.secho("Setting up you Youtube configuration...", fg="magenta")
    click.echo()
    api_key = click.prompt("Enter your API key")
    username = click.prompt("What would you name this key?")

    config_file_path = Path(_get_config_path())

    if not config_file_path.parent.exists():
        click.echo()
        click.secho(
            "Creating config folder at %s" % config_file_path.parent, fg="magenta"
        )
        config_file_path.parent.mkdir(parents=True)

    conf = _set_up_config()
    click.echo(f"Config file is stored at {config_file_path.resolve()}.")

    try:
        conf.add_profile(name=username, key=api_key)
    except ValueAlreadyExists:
        raise click.BadArgumentUsage("API key already exists in config file.")

    click.echo()
    if click.confirm("Set this API key as default?"):
        conf.set_default(username)
    click.echo()
    click.secho("API key successfully configured!", fg="green", bold=True)
    click.echo()
    click.echo("To add more API keys, rerun `youte config add-key`.")
    click.echo("To set a default key, run `youte config set-default <name-of-key>`")


@config.command()
@click.argument("name")
def set_default(name):
    """Set default API key"""
    click.echo("Setting %s as default key" % name)
    conf = _set_up_config()
    try:
        conf.set_default(name)
        click.echo("%s is now your default API key" % conf[name]["key"])
    except KeyError:
        click.secho(
            """No such name found.
            You might not have added this profile or
            added it under a different name.
            Run:
            - `youte config list-keys` to see the list of registered keys.
            - `youte config add-key to add a new API key.
            """,
            fg="red",
            bold=True,
        )


@config.command()
@click.argument("name")
def remove(name):
    """Delete stored API key"""
    conf = _set_up_config()
    conf.delete_profile(name=name)
    click.echo(f"Profile `{name}` successfully removed!")


@config.command()
def list_keys():
    """Show a list of keys already added"""
    click.echo()
    conf = _set_up_config()

    if not Path(get_config_path()).exists():
        click.echo("No API key has been added to config file.")
    else:
        for name in conf:
            if "default" in conf[name]:
                click.secho("%s -------- %s (*)" % (name, conf[name]["key"]), fg="blue")
            else:
                click.echo("%s -------- %s" % (name, conf[name]["key"]))
        click.echo()
        click.secho(
            "All API keys are stored in %s" % get_config_path(), fg="green", bold=True
        )
        click.echo()


def _set_up_collector(api_key: str) -> Youte:
    collector = Youte(api_key=api_key)

    return collector


def _get_config_path(filename="config"):
    config_file_path = Path(click.get_app_dir("youte")).joinpath(filename)
    return str(config_file_path)


def _set_up_config(filename=_get_config_path()) -> YouteConfig:
    config_file = YouteConfig(filename=filename)

    return config_file


def _get_ids(string: Sequence[str] = None, file: str = None) -> List[str]:
    _raise_no_item_error(items=string, file_path=file)

    if string:
        ids = list(string)

    if file:
        with open(file, mode="r") as f:
            ids = [row.rstrip() for row in f.readlines()]

    return ids


def _raise_no_item_error(items: Sequence[str], file_path: str) -> None:
    if not (items or file_path):
        click.secho(
            "No item ID is specified. Pass some item IDs or specify a file.",
            fg="red",
            bold=True,
        )
        sys.exit(1)


def _prompt_save_progress(filename) -> None:
    if click.confirm("Do you want to save your current progress?"):
        full_path = Path(filename).resolve()
        click.echo(f"Progress saved at {full_path}")
        click.echo(
            "To resume progress, run the same youte search command "
            f"and add `--resume {full_path.stem}`"
        )
    else:
        try:
            os.remove(filename)
        except FileNotFoundError:
            pass
    sys.exit(0)


def _get_api_key(name=None, filename="config"):
    """Get API key from config file.
    If no name is given, use default API key
    """
    logger.info("Getting API key from config file.")
    config_file_path = Path(click.get_app_dir("youte")).joinpath(filename)
    config = YouteConfig(filename=str(config_file_path))

    if name:
        try:
            api_key = config[name]["key"]
        except KeyError:
            click.secho("ERROR", fg="red", bold=True)
            click.secho(
                "No API key found for %s. Did you use a different name?\n"
                "Try:\n"
                "`youte config list-keys` to get a "
                "full list of registered API keys "
                "or `youte config add-key` to add a new API key" % name,
                fg="red",
                bold=True,
            )
            sys.exit(1)
    else:
        default_check = []
        for name in config:
            if "default" in config[name]:
                default_check.append(name)
        if not default_check:
            click.secho(
                "No API key name was specified, and you haven't got a default API key.",
                fg="red",
                bold=True,
            )
            sys.exit(1)
        else:
            api_key = config[default_check[0]]["key"]

    return api_key


